ownCloud/Migration/Migration.swift

```diff
@@ -0,0 +1,517 @@
+	}
+
+	var title: String?
+	var description: String?
+	var state: State = .initiated
+	var type: ActivityType = .account
+}
+
+struct Entitlements : Codable {
+
+	var appGroups : [String]?
+	var keychainAccessGroups : [String]?
+
+	enum CodingKeys: String, CodingKey {
+		case appGroups = "com.apple.security.application-groups"
+		case keychainAccessGroups = "keychain-access-groups"
+	}
+}
+
+class Migration {
+
+	static let ActivityUpdateNotification = NSNotification.Name(rawValue: "MigrationActivityUpdateNotification")
+	static let FinishedNotification = NSNotification.Name(rawValue: "MigrationFinishedNotification")
+
+	private static let legacyDbFilename = "DB.sqlite"
+	private static let legacyCacheFolder = "cache_folder"
+	private static let legacyInstantUploadFolder = "/InstantUpload"
+
+	static let shared = Migration()
+
+	private lazy var appGroupId : String? = {
+		// Try to read entitlements
+		if let entitlements = readAppEntitlements() {
+			return entitlements.appGroups?.first
+		}
+
+		// Try to construct group id from bundle identifier
+		if let bundleId = Bundle.main.bundleIdentifier {
+			return "group." + bundleId
+		}
+
+		return nil
+	}()
+
+	private lazy var legacyDataDirectoryURL: URL? = {
+		if let groupId = self.appGroupId {
+			let containerURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: groupId)
+			return containerURL?.appendingPathComponent(Migration.legacyCacheFolder)
+		}
+		return nil
+	}()
+
+	// MARK: - Public API
+
+	var legacyDataFound : Bool {
+		var isDirectory : ObjCBool = false
+		if let directoryPath = self.legacyDataDirectoryURL?.path {
+			let pathExists = FileManager.default.fileExists(atPath: directoryPath, isDirectory: &isDirectory)
+			return (pathExists && isDirectory.boolValue == true)
+		}
+		return false
+	}
+
+	private let migrationQueue = DispatchQueue(label: "com.owncloud.migration-queue")
+
+	func migrateAccountsAndSettings(_ parentViewController:UIViewController? = nil) {
+
+		guard let legacyDbURL = self.legacyDataDirectoryURL?.appendingPathComponent(Migration.legacyDbFilename) else { return }
+
+		if FileManager.default.fileExists(atPath: legacyDbURL.path) {
+			let db = OCSQLiteDB(url: legacyDbURL)
+			db.open(with: .readOnly) { (_, err) in
+				if err == nil {
+					Log.debug(tagged: ["MIGRATION"], "Legacy database successfully opened")
+					let queryResultHandler : OCSQLiteDBResultHandler = { (db, error, transaction, resultSet) in
+
+						if error != nil {
+							Log.error(tagged: ["MIGRATION"], "Failed to fetch users table from legacy database with error: \(String(describing: error))")
+						}
+
+						if let resultSet = resultSet {
+							var error : NSError?
+
+							resultSet.iterate({ (_, _, rowDict, _) in
+								self.migrationQueue.async {
+									if let userId = rowDict["id"] as? Int,
+										let serverURL = rowDict["url"] as? String,
+										let credentials = self.getCredentialsDataItem(for: userId) {
```

Comment:
For reasons unbeknownst to me (app prefix identifier, group ID and keychain entry name seem to match in the debugger), the new app trying to access the credentials from the legacy app (whose bundleID I changed to `com.owncloud.ios-app`) fails at `SecItemCopyMatching` with a `errSecItemNotFound` return value, resulting in a `nil` value being returned from `getCredentialsDataItem`.

The `if` condition above subsequently fails, leading to this blank view:
<img width="386" alt="Bildschirmfoto 2020-05-11 um 23 24 19" src="https://user-images.githubusercontent.com/639669/81613316-8b75d700-93de-11ea-9361-ec7818799eeb.png">

Reply:
@felix-schwarz I will revisit this part and fix the handling

Reply:
@felix-schwarz Fixed it, now the bookmark migration will be shown as failed in case no data can't be retrieved from keychain. The question is however: should we allow migrating accounts for which no authorisation data can't be read still? And just show a warning that user will have to re-authorize this bookmark?

Reply:
@mneuwert In the wild, accounts for which no auth data can be read should be relatively rare. Although, of course, they can naturally occur as a byproduct if someone restores their iPhone from an unencrypted backup (restoring the app's data, but not its keychain).

If it's not too much effort, handling this scenario as well would of course be nice. But it clearly is a niche case.

Reply:
@felix-schwarz Actually my hope to do something about such account has vanished after I double-checked the code.. Legacy app has only stored URL in the database and user ID (which is just a counter) which referenced data blob in the keychain (storing username, auth method, auth token etc). So If keychain is lost, URL IMO is not sufficient.